src/merkle/mmr/partial.rs

```diff
@@ -0,0 +1,362 @@
+    ///    elements are tracked.
+    /// 2. During a [Mmr] update peaks can be merged by hashing the left and right hand sides. The
+    ///    peaks are used as the left hand.
+    ///
+    /// All the peaks of every tree in the [Mmr] forest. The peaks are always ordered by number of
+    /// leaves, starting from the peak with most children, to the one with least.
+    pub(crate) peaks: Vec<RpoDigest>,
+
+    /// Authentication nodes used to construct merkle paths for a subset of the [Mmr]'s leaves.
+    ///
+    /// This does not include the [Mmr]'s peaks nor the tracked nodes, only the elements required
+    /// to construct their authentication paths. This property is used to detect when elements can
+    /// be safely removed from, because they are no longer required to authenticate any element in
+    /// the [PartialMmr].
+    ///
+    /// The elements in the [Mmr] are referenced using a in-order tree index. This indexing scheme
+    /// permits for easy computation of the relative nodex (left/right children, sibling, parent),
+    /// which is useful for traversal. The indexing is also stable, meaning that merges to the
+    /// trees in the [Mmr] can be represented without rewrites of the indexes.
+    pub(crate) nodes: BTreeMap<InOrderIndex, RpoDigest>,
+
+    /// Flag indicating if the odd element should be tracked.
+    ///
+    /// This flag is necessary because the sibling of the odd doesn't exist yet, so it can not be
+    /// added into `nodes` to signal the value is being tracked.
+    pub(crate) track_latest: bool,
+}
+
+impl PartialMmr {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Constructs a [PartialMmr] from the given [MmrPeaks].
+    pub fn from_peaks(accumulator: MmrPeaks) -> Self {
+        let forest = accumulator.num_leaves();
+        let peaks = accumulator.peaks().clone();
+        let nodes = BTreeMap::new();
+        let track_latest = false;
+
+        Self { forest, peaks, nodes, track_latest }
+    }
+
+    // ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    // Gets the current `forest`.
+    //
+    // This value corresponds to the version of the [PartialMmr] and the number of leaves in it.
+    pub fn forest(&self) -> usize {
+        self.forest
+    }
+
+    // Returns a reference to the current peaks in the [PartialMmr]
+    pub fn peaks(&self) -> &[RpoDigest] {
+        &self.peaks
+    }
+
+    /// Return the authentication path of `leaf_pos`.
+    ///
+    /// This method returns an error if `leaf_pos` is too large, and `None` if the requested value
+    /// is not tracked.
+    ///
+    /// Note: `leaf_pos` corresponds to the position the [Mmr] and not on an individual tree.
+    pub fn get_merkle_path(&self, leaf_pos: usize) -> Result<Option<MerklePath>, MmrError> {
```

Comment:
Yes. `Result` is used to signal an error state, `Option` is signal to signal the lack of a proof. That is not an error state but a normal usage of the structure.

I guess one could encode the state in the `MmrError` but then one loses the ability of using the operator `?` on the result.

Reply:
Sorry, I wasn't clear here. We should still return the result. My suggestion was to replace `MmrProof` rather than `MerklePath`. Basically, the full return type would be `Result<Option<MmrProof>, MmrError>`.

Reply:
Oh, sorry, I misunderstood the request. Changing the return type, and the name of the method to `open` so that it matches the `Mmr`.